Enable mllm api

xuechendi opened the pull request:
Purpose of the PR:
Enable MLLM model support in LLM-on-Ray: OpenAI compatible AI support, webui support

Models:
Any huggingface MLLM models, tested with below two
  - Fuyu-8b: https://huggingface.co/adept/fuyu-8b
  - Deplot: https://huggingface.co/google/deplot

Config update:
![image](https://github.com/intel/llm-on-ray/assets/4355494/e20d0c25-b2d0-46b6-8b37-2e3877e62e97)

UI example:
![image](https://github.com/intel/llm-on-ray/assets/4355494/b7892128-44b9-47c6-86d1-fbc616a8fa1c)
![image](https://github.com/intel/llm-on-ray/assets/4355494/fc5a0314-5fe0-471d-b5af-1f5fa39670c0)

carsonwang commented on ui/start_ui.py:
```diff
@@ -674,10 +758,14 @@ def shutdown_deploy(self):
 
     def get_ray_cluster(self):
-        command = "conda activate " + self.conda_env_name + "; ray status"
+        command = "source ~/anaconda3/bin/activate; conda activate " + self.conda_env_name + "; ray status"
```
We can't assume this path works in other users' environment.

xuechendi replied:
Sorry, Forgot to remove that one.

xuechendi replied:
fixed

carsonwang reviewed (commented):
Thank you @xuechendi for the work! Can we also add one of the models in the CI?

xuechendi commented on inference/predictor_deployment.py:
```diff
@@ -166,19 +176,31 @@ async def openai_call(self, prompt, config, streaming_response=True):
+                generate_text = generate_result[0]
+                model_response = ModelResponse(
+                    generated_text=generate_text,
+                    num_input_tokens=self.predictor.input_length,
```
The `generate_result` is type as "String" when I use my test model `Fuyu-8b`, so it does not have `input_length` properties like other predictors.

xuechendi replied:
> The `generate_result` is type as "String" when I use my test model `Fuyu-8b`, so it does not have `input_length` properties like other predictors.

Just realized that `GenerateResult` is instantiate in transformer_predictor, in that case, I'll merge here

xuechendi commented:
@carsonwang , I fixed issues per your request. 
1) change to use URL in my example
2) update script description for my example
3) removed conda assumption, fix typo bug in prediction_deployment.py
4) update MLLM_predictor.py with Ipex support, torch_dtype, add GenerateResult wrapper class

xuechendi commented on examples/inference/api_server_openai/image_query_http_requests.py:
```diff
@@ -0,0 +1,131 @@
+)
+parser.add_argument(
+    "--image_path",
+    default="my_app/test_data/twitter_graph.png",
```
fixed

xuechendi commented on examples/inference/api_server_openai/image_query_http_requests.py:
```diff
@@ -0,0 +1,131 @@
+    os.environ['WORKDIR'] = os.getcwd()
+
+parser = argparse.ArgumentParser(
+    description="Example script to query with http requests", add_help=True
```
fixed

xuechendi commented on inference/mllm_predictor.py:
```diff
@@ -0,0 +1,72 @@
+
+            adapt_transformers_to_gaudi()
+        # get correct torch type for loading HF model
+        # torch_dtype = get_torch_dtype(infer_conf, hf_config)
```
fixed

xuechendi commented on inference/mllm_predictor.py:
```diff
@@ -0,0 +1,72 @@
+        # torch_dtype = get_torch_dtype(infer_conf, hf_config)
+        
+        # model = FuyuForCausalLM.from_pretrained(model_desc.model_id_or_path)
+        # processor = FuyuProcessor.from_pretrained(model_desc.model_id_or_path)
```
fixed

xuechendi commented on inference/models/fuyu8b.yaml:
```diff
@@ -0,0 +1,26 @@
+  precision: bf16
+model_description:
+  model_id_or_path: /mnt/nvme0n1/chendi/llm-on-ray/models/adept/fuyu-8b
+  tokenizer_name_or_path: /mnt/nvme0n1/chendi/llm-on-ray/models/adept/fuyu-8b
```
fixed

carsonwang commented on examples/inference/api_server_openai/query_http_requests.py:
```diff
@@ -35,17 +35,21 @@
 parser.add_argument(
     "--top_p",
-    default=None,
+    default=0.7,
```
Can we revoke the default value changes in this file? If these default values work for the mllm models, it makes sense to set them in `image_query_http_requests.py` as you have done. But this file more general so let's just use openai's default value?

carsonwang commented on examples/inference/api_server_openai/query_openai_sdk.py:
```diff
@@ -41,21 +41,47 @@
+        stream=True,
+        max_tokens=128,
+        temperature=0.2,
+        top_p=0.7,
```
Use args.xxx

carsonwang commented on examples/inference/api_server_openai/query_openai_sdk.py:
```diff
@@ -41,21 +41,47 @@
+        stream=False,
+        max_tokens=128,
+        temperature=0.2,
+        top_p=0.7,
```
Use args.xxx

carsonwang reviewed (commented):
This looks mostly good. @KepingYan please take another look especially for the UI changes.

KepingYan commented on ui/start_ui.py:
```diff
@@ -1024,22 +1134,32 @@ def _init_ui(self):
+                        with gr.Accordion("Image", open=False, visible=True):
+                            image = gr.Image(type="pil")
+                    with gr.Row():
+                        endpoint_value = "http://127.0.0.1:8000/v1/chat/completions"
+                        model_endpoint = gr.Text(
+                            label="Model Endpoint", value=endpoint_value, scale=1
+                        )
```
Is it better for the default value to be None? We can set placeholder to let users know that it can be a pre-deployed endpoint or the endpoint returned from deployment module.

KepingYan commented on ui/start_ui.py:
```diff
@@ -1024,22 +1134,32 @@ def _init_ui(self):
+                            label="Model Endpoint", value=endpoint_value, scale=1
+                        )
+                        model_name = gr.Text(
+                            label="Model Name", value="llama-2-7b-chat-hf", scale=1
```
Same here. Because the default value may not be deployed by users.

xuechendi commented:
@KepingYan  @carsonwang , I completed all codes fixing per your review. Thanks

KepingYan commented on examples/inference/api_server_openai/query_openai_sdk.py:
```diff
@@ -41,21 +41,47 @@
-    top_p=args.top_p,
-)
-print(chat_completion)
+client = OpenAI(base_url="http://localhost:8000/v1", api_key="not_needed")
```
We can remove parameters here, and set it via environment variables OPENAI_BASE_URL and OPENAI_API_KEY as described in readme.

KepingYan commented on examples/inference/api_server_openai/query_openai_sdk.py:
```diff
@@ -41,21 +41,47 @@
+        temperature=args.temperature,
+        top_p=args.top_p,
+    ):
+        content = chunk.choices[0].delta.content
```
Is it necessary to verify content of the chunk here? such as chunk != None and chunk !=""

xuechendi commented:
@KepingYan , I updated 4 examples by using ENV variables to set base url and api_key
1. langchain_sdk for image/text
2. openai sdk for image/text

carsonwang reviewed (approved):
Thank you @xuechendi ！ Merging this.
